src/ethereum/forks/amsterdam/vm/instructions/system.py

```diff
@@ -654,12 +691,22 @@ def delegatecall(evm: Evm) -> None:
         code,
         delegated_access_gas_cost,
     ) = access_delegation(evm, code_address)
-    access_gas_cost += delegated_access_gas_cost
 
+    # Charge delegation access cost before tracking delegation target
+    if delegated_access_gas_cost > Uint(0):
+        charge_gas(evm, delegated_access_gas_cost)
+
+    # Track delegation target if delegation occurred
+    if disable_precompiles:
+        track_address_access(
+            evm.message.block_env.state.change_tracker, code_address
+        )
+
+    # Charge remaining costs (subcall)
     message_call_gas = calculate_message_call_gas(
-        U256(0), gas, Uint(evm.gas_left), extend_memory.cost, access_gas_cost
+        U256(0), gas, Uint(evm.gas_left), Uint(0), Uint(0)
```

Comment:
I think this kind of thing is only relevant now with BAL or state tracking in general but this should likely be reflected back to all forks right? There are things in BAL implementation that may impact the logic across all forks.

I think this plays into the state tracker conversation where this would have to be refactored for all forks. And it sort of renders memory and access costs in `calculate_message_call_gas` unnecessary (I believe).